### PR TITLE
Implement detach method

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/3074
## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/file/service/endpoint/Stop.java [hunk 1]
```diff
@@ -31,12 +31,14 @@
...
     public static Object stop(BObject listener) {
-        LocalFileSystemServerConnector serverConnector = (LocalFileSystemServerConnector) listener
-                .getNativeData(DirectoryListenerConstants.FS_SERVER_CONNECTOR);
-        try {
-            serverConnector.stop();
-        } catch (LocalFileSystemServerConnectorException e) {
-            return FileUtils.getBallerinaError(FileConstants.FILE_SYSTEM_ERROR, e.getMessage());
+        if (listener.getNativeData(DirectoryListenerConstants.FS_SERVER_CONNECTOR) != null) {
+            LocalFileSystemServerConnector serverConnector = (LocalFileSystemServerConnector) listener
+                    .getNativeData(DirectoryListenerConstants.FS_SERVER_CONNECTOR);
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/endpoint/Stop.java:36]:
```suggestion
        Object fsServerConnector = listener.getNativeData(DirectoryListenerConstants.FS_SERVER_CONNECTOR);
        if (fsServerConnector instanceof LocalFileSystemServerConnector) {
            LocalFileSystemServerConnector localServerConnector = (LocalFileSystemServerConnector) fsServerConnector;
```

File: ballerina/tests/directory-listener-test.bal [hunk 2]
```diff
@@ -165,6 +165,18 @@ function testService() returns error? {
...
 
+@test:Config {}
+function testService1() returns error? {
+    Listener|error temporaryLoader = localFolder5;
+    if temporaryLoader is Listener {
+        check temporaryLoader.attach(attachService1);
+        check temporaryLoader.'start();
+        check temporaryLoader.immediateStop();
+    } else {
+        test:assertFail("Test Failed!");
```
Comment (MadhukaHarith92) [inline ballerina/tests/directory-listener-test.bal:176]:
```suggestion
        test:assertFail("The temporaryLoader is not a Listener");
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 3]
```diff
@@ -46,36 +47,33 @@ public class FSListener implements LocalFileSystemListener {
 
     private static final Logger log = LoggerFactory.getLogger(FSListener.class);
     private Runtime runtime;
-    private BObject service;
-    private Map<String, MethodType> attachedFunctionRegistry;
+    private Map<BObject, Map<String, MethodType>> attachedFunctionRegistries = new HashMap<>();
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:50]:
```suggestion
    private Map<BObject, Map<String, MethodType>> serviceRegistry = new HashMap<>();
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 4]
```diff
@@ -46,36 +47,33 @@ public class FSListener implements LocalFileSystemListener {
...
-            if (service.getType().isIsolated()
-                    && service.getType().isIsolated(resourceName)) {
-                runtime.invokeMethodAsyncConcurrently(service, resourceName, null,
-                        ON_MESSAGE_METADATA, new DirectoryCallback(), null, PredefinedTypes.TYPE_NULL,
-                        parameters);
-            } else {
-                runtime.invokeMethodAsyncSequentially(service, resourceName, null,
-                        ON_MESSAGE_METADATA, new DirectoryCallback(), null, PredefinedTypes.TYPE_NULL,
-                        parameters);
+        for (Map.Entry<BObject, Map<String, MethodType>> attachedFunctionRegistry:
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:61]:
```suggestion
        for (Map.Entry<BObject, Map<String, MethodType>> serviceEntry:
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 5]
```diff
@@ -88,7 +86,15 @@ private Object[] getJvmSignatureParameters(LocalFileSystemEvent fileEvent) {
         return new Object[] { eventStruct, true };
     }
 
-    private MethodType getMethodType(String event) {
+    private MethodType getMethodType(Map<String, MethodType> attachedFunctionRegistry, String event) {
         return attachedFunctionRegistry.get(event);
     }
+
+    public void addService(BObject service, Map<String, MethodType> attachedFunctionRegistry) {
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:93]:
```suggestion
    public void addService(BObject service, Map<String, MethodType> attachedFunctions) {
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 6]
```diff
@@ -46,36 +47,33 @@ public class FSListener implements LocalFileSystemListener {
...
-                runtime.invokeMethodAsyncConcurrently(service, resourceName, null,
-                        ON_MESSAGE_METADATA, new DirectoryCallback(), null, PredefinedTypes.TYPE_NULL,
-                        parameters);
-            } else {
-                runtime.invokeMethodAsyncSequentially(service, resourceName, null,
-                        ON_MESSAGE_METADATA, new DirectoryCallback(), null, PredefinedTypes.TYPE_NULL,
-                        parameters);
+        for (Map.Entry<BObject, Map<String, MethodType>> attachedFunctionRegistry:
+                attachedFunctionRegistries.entrySet()) {
+            MethodType resource = getMethodType(attachedFunctionRegistry.getValue(), fileEvent.getEvent());
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:63]:
```suggestion
            MethodType serviceFunction = serviceEntry.getValue().get(fileEvent.getEvent());
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 7]
```diff
@@ -46,36 +47,33 @@ public class FSListener implements LocalFileSystemListener {
...
+            if (resource != null) {
+                String resourceName = resource.getName();
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:65]:
```suggestion
                String functionName = resource.getName();
```

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java [hunk 8]
```diff
@@ -88,7 +86,15 @@ private Object[] getJvmSignatureParameters(LocalFileSystemEvent fileEvent) {
         return new Object[] { eventStruct, true };
     }
 
-    private MethodType getMethodType(String event) {
+    private MethodType getMethodType(Map<String, MethodType> attachedFunctionRegistry, String event) {
         return attachedFunctionRegistry.get(event);
     }
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java:91]: we don't need a method ?

File: native/src/main/java/io/ballerina/stdlib/file/service/endpoint/InitEndpoint.java [hunk 9]
```diff
@@ -46,8 +55,33 @@ public static Object initEndpoint(BObject listener) {
...
+                    new FSListener(env.getRuntime()));
+            listener.addNativeData(DirectoryListenerConstants.FS_SERVER_CONNECTOR, serverConnector);
+        } catch (LocalFileSystemServerConnectorException e) {
+            return FileUtils.getBallerinaError(FileConstants.FILE_SYSTEM_ERROR,
+                    "Unable to initialize server connector: " + e.getMessage());
+        }
         return null;
     }
 
+    private static Map<String, String> getParamMap(BMap serviceEndpointConfig) {
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/endpoint/InitEndpoint.java:73]:
```suggestion
    private static Map<String, String> readConfiguration(BMap serviceEndpointConfig) {
```

File: native/src/main/java/io/ballerina/stdlib/file/service/endpoint/InitEndpoint.java [hunk 10]
```diff
@@ -46,8 +55,33 @@ public static Object initEndpoint(BObject listener) {
         if (!Files.isDirectory(dirPath)) {
             return FileUtils.getBallerinaError(FileConstants.FILE_SYSTEM_ERROR, "Unable to find a directory: " + path);
         }
+        FileSystemConnectorFactory connectorFactory = new FileSystemConnectorFactoryImpl();
+
+        final Map<String, String> paramMap = getParamMap(serviceEndpointConfig);
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/file/service/endpoint/InitEndpoint.java:60]:
```suggestion
        final Map<String, String> configMap = readConfiguation(serviceEndpointConfig);
```

**Comment by daneshk** [comment]
@kalaiyarasiganeshalingam Add a few comments, especially to give meaningful names for variables and functions. can you check and change

File: native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java (hunk 8 above)
Comment (kalaiyarasiganeshalingam) [reply native/src/main/java/io/ballerina/stdlib/file/service/FSListener.java]: Yes

**Comment by kalaiyarasiganeshalingam** [comment]
> @kalaiyarasiganeshalingam Add a few comments, especially to give meaningful names for variables and functions. can you check and change

Okay. Yes.

Fixed now.

**Review by daneshk — APPROVED** [review]
LGTM. we may have to wait till update 2 release

**Pull request merged**